File: backend/app/routers/chats.py
import json
from typing import AsyncIterator

from fastapi import APIRouter, Depends, HTTPException
from fastapi.responses import StreamingResponse
from sqlalchemy.orm import Session, joinedload

from app.config import settings
from app.database import SessionLocal, get_db
from app.models import (
    Chat,
    Character,
    Fact,
    GenerationJob,
    JobStatus,
    Message,
    MessageRole,
)
from app.queue import enqueue_generation
from app.schemas import (
    ChatCreate,
    ChatOut,
    ChatUpdate,
    FactOut,
    FactUpdate,
    JobOut,
    MessageOut,
    QueueStatus,
    SendMessageRequest,
)
from app.settings_service import get_or_create_settings

router = APIRouter(prefix="/chats", tags=["chats"])


def _job_out(job: GenerationJob) -> JobOut:
    return JobOut.model_validate(job)


def _chat_out(db: Session, chat: Chat) -> ChatOut:
    active = (
        db.query(GenerationJob)
        .filter(
            GenerationJob.chat_id == chat.id,
            GenerationJob.status.in_([JobStatus.queued, JobStatus.running]),
        )
        .order_by(GenerationJob.created_at.asc())
        .first()
    )
    queued = (
        db.query(GenerationJob)
        .filter(
            GenerationJob.chat_id == chat.id,
            GenerationJob.status == JobStatus.queued,
        )
        .count()
    )
    return ChatOut(
        id=chat.id,
        title=chat.title,
        character_id=chat.character_id,
        summary=chat.summary,
        created_at=chat.created_at,
        updated_at=chat.updated_at,
        active_job=_job_out(active) if active else None,
        queued_jobs=queued,
    )


@router.get("", response_model=list[ChatOut])
def list_chats(db: Session = Depends(get_db)):
    chats = db.query(Chat).order_by(Chat.updated_at.desc()).all()
    return [_chat_out(db, chat) for chat in chats]


@router.post("", response_model=ChatOut)
def create_chat(payload: ChatCreate, db: Session = Depends(get_db)):
    if payload.character_id:
        character = (
            db.query(Character).filter(Character.id == payload.character_id).first()
        )
        if not character:
            raise HTTPException(status_code=404, detail="Character not found")

    chat = Chat(title=payload.title, character_id=payload.character_id)
    db.add(chat)
    db.commit()
    db.refresh(chat)

    if payload.character_id:
        character = (
            db.query(Character).filter(Character.id == payload.character_id).first()
        )
        if character and character.first_message.strip():
            db.add(
                Message(
                    chat_id=chat.id,
                    role=MessageRole.assistant,
                    content=character.first_message.strip(),
                )
            )
            db.commit()

    return _chat_out(db, chat)


@router.get("/queue", response_model=QueueStatus)
def get_queue(db: Session = Depends(get_db)):
    running = (
        db.query(GenerationJob)
        .filter(GenerationJob.status == JobStatus.running)
        .order_by(GenerationJob.started_at.asc())
        .all()
    )
    queued = (
        db.query(GenerationJob)
        .filter(GenerationJob.status == JobStatus.queued)
        .order_by(GenerationJob.created_at.asc())
        .all()
    )
    return QueueStatus(
        running=[_job_out(j) for j in running],
        queued=[_job_out(j) for j in queued],
    )


@router.get("/{chat_id}", response_model=ChatOut)
def get_chat(chat_id: int, db: Session = Depends(get_db)):
    chat = db.query(Chat).filter(Chat.id == chat_id).first()
    if not chat:
        raise HTTPException(status_code=404, detail="Chat not found")
    return _chat_out(db, chat)


@router.patch("/{chat_id}", response_model=ChatOut)
def update_chat(chat_id: int, payload: ChatUpdate, db: Session = Depends(get_db)):
    chat = db.query(Chat).filter(Chat.id == chat_id).first()
    if not chat:
        raise HTTPException(status_code=404, detail="Chat not found")
    for key, value in payload.model_dump(exclude_unset=True).items():
        setattr(chat, key, value)
    db.commit()
    db.refresh(chat)
    return _chat_out(db, chat)


@router.delete("/{chat_id}")
def delete_chat(chat_id: int, db: Session = Depends(get_db)):
    chat = db.query(Chat).filter(Chat.id == chat_id).first()
    if not chat:
        raise HTTPException(status_code=404, detail="Chat not found")
    db.delete(chat)
    db.commit()
    return {"ok": True}


@router.get("/{chat_id}/messages", response_model=list[MessageOut])
def list_messages(chat_id: int, db: Session = Depends(get_db)):
    chat = db.query(Chat).filter(Chat.id == chat_id).first()
    if not chat:
        raise HTTPException(status_code=404, detail="Chat not found")

    messages = (
        db.query(Message)
        .options(joinedload(Message.job))
        .filter(Message.chat_id == chat_id)
        .order_by(Message.created_at.asc())
        .all()
    )
    result: list[MessageOut] = []
    for msg in messages:
        job_status = msg.job.status if msg.job else None
        result.append(
            MessageOut(
                id=msg.id,
                chat_id=msg.chat_id,
                role=msg.role,
                content=msg.content,
                is_summary=msg.is_summary,
                created_at=msg.created_at,
                job_status=job_status,
            )
        )
    return result


@router.post("/{chat_id}/messages", response_model=MessageOut)
def send_message(
    chat_id: int, payload: SendMessageRequest, db: Session = Depends(get_db)
):
    chat = db.query(Chat).filter(Chat.id == chat_id).first()
    if not chat:
        raise HTTPException(status_code=404, detail="Chat not found")
    if not payload.content.strip():
        raise HTTPException(status_code=400, detail="Message cannot be empty")

    get_or_create_settings(db)

    user_msg = Message(
        chat_id=chat_id,
        role=MessageRole.user,
        content=payload.content.strip(),
    )
    db.add(user_msg)

    assistant_msg = Message(
        chat_id=chat_id,
        role=MessageRole.assistant,
        content="",
    )
    db.add(assistant_msg)
    db.commit()
    db.refresh(assistant_msg)

    job = enqueue_generation(db, chat_id, assistant_msg.id)
    chat.updated_at = assistant_msg.created_at
    db.commit()

    return MessageOut(
        id=assistant_msg.id,
        chat_id=assistant_msg.chat_id,
        role=assistant_msg.role,
        content=assistant_msg.content,
        is_summary=assistant_msg.is_summary,
        created_at=assistant_msg.created_at,
        job_status=job.status,
    )


@router.get("/{chat_id}/facts", response_model=list[FactOut])
def list_facts(chat_id: int, db: Session = Depends(get_db)):
    chat = db.query(Chat).filter(Chat.id == chat_id).first()
    if not chat:
        raise HTTPException(status_code=404, detail="Chat not found")
    return (
        db.query(Fact).filter(Fact.chat_id == chat_id).order_by(Fact.key.asc()).all()
    )


@router.put("/{chat_id}/facts", response_model=FactOut)
def upsert_fact(chat_id: int, payload: FactUpdate, db: Session = Depends(get_db)):
    chat = db.query(Chat).filter(Chat.id == chat_id).first()
    if not chat:
        raise HTTPException(status_code=404, detail="Chat not found")

    fact = (
        db.query(Fact)
        .filter(Fact.chat_id == chat_id, Fact.key == payload.key)
        .first()
    )
    if fact:
        fact.value = payload.value
    else:
        fact = Fact(chat_id=chat_id, key=payload.key, value=payload.value)
        db.add(fact)
    db.commit()
    db.refresh(fact)
    return fact


@router.delete("/{chat_id}/facts/{fact_key}")
def delete_fact(chat_id: int, fact_key: str, db: Session = Depends(get_db)):
    fact = (
        db.query(Fact).filter(Fact.chat_id == chat_id, Fact.key == fact_key).first()
    )
    if not fact:
        raise HTTPException(status_code=404, detail="Fact not found")
    db.delete(fact)
    db.commit()
    return {"ok": True}


async def _chat_event_stream(chat_id: int) -> AsyncIterator[str]:
    last_payload = ""
    while True:
        db = SessionLocal()
        try:
            chat = db.query(Chat).filter(Chat.id == chat_id).first()
            if not chat:
                yield f"event: error\ndata: {json.dumps({'detail': 'not found'})}\n\n"
                break

            messages = (
                db.query(Message)
                .options(joinedload(Message.job))
                .filter(Message.chat_id == chat_id)
                .order_by(Message.created_at.asc())
                .all()
            )
            active_job = (
                db.query(GenerationJob)
                .filter(
                    GenerationJob.chat_id == chat_id,
                    GenerationJob.status.in_(
                        [JobStatus.queued, JobStatus.running]
                    ),
                )
                .order_by(GenerationJob.created_at.desc())
                .first()
            )

            payload = json.dumps(
                {
                    "chat": _chat_out(db, chat).model_dump(mode="json"),
                    "messages": [
                        MessageOut(
                            id=m.id,
                            chat_id=m.chat_id,
                            role=m.role,
                            content=m.content,
                            is_summary=m.is_summary,
                            created_at=m.created_at,
                            job_status=m.job.status if m.job else None,
                        ).model_dump(mode="json")
                        for m in messages
                    ],
                    "active_job": _job_out(active_job).model_dump(mode="json")
                    if active_job
                    else None,
                },
                default=str,
            )

            if payload != last_payload:
                last_payload = payload
                yield f"data: {payload}\n\n"

            if not active_job:
                yield f"event: idle\ndata: {json.dumps({'chat_id': chat_id})}\n\n"
                break
        finally:
            db.close()

        import asyncio

        await asyncio.sleep(settings.sse_poll_interval)


@router.get("/{chat_id}/stream")
async def stream_chat(chat_id: int, db: Session = Depends(get_db)):
    chat = db.query(Chat).filter(Chat.id == chat_id).first()
    if not chat:
        raise HTTPException(status_code=404, detail="Chat not found")
    return StreamingResponse(
        _chat_event_stream(chat_id),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )

File: crates/server/src/routes/chats.rs
use std::convert::Infallible;
use std::time::Duration;

use async_stream::stream;
use axum::{
    extract::{Path, State},
    response::{
        sse::{Event, KeepAlive, Sse},
        IntoResponse,
    },
    routing::{delete, get, patch, post},
    Json, Router,
};
use dreamwell_types::{
    Chat, ChatCreate, ChatStreamPayload, ChatUpdate, ChatVariable, ChatVariableUpdate, Job,
    Message, MessageRole, OkResponse, QueueStatus, RegenerateMessageRequest,
    RegenerateSummaryRequest, SendMessageRequest, UpdateMessageRequest,
};

use crate::db;
use crate::error::{AppError, AppResult};
use crate::queue::enqueue_generation;
use crate::routes::AppState;
use crate::variables::{
    revert_message_variable_updates, revert_variable_updates_from_messages,
    strip_variable_key_from_chat_messages,
};

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/queue", get(get_queue))
        .route("/archived", get(list_archived_chats))
        .route("/", get(list_chats).post(create_chat))
        .route(
            "/:id",
            get(get_chat).patch(update_chat).delete(archive_chat),
        )
        .route("/:id/restore", post(restore_chat))
        .route("/:id/permanent", delete(permanently_delete_chat))
        .route("/:id/messages", get(list_messages).post(send_message))
        .route(
            "/:id/messages/:message_id",
            patch(update_message).delete(rewind_message),
        )
        .route(
            "/:id/messages/:message_id/regenerate",
            post(regenerate_message),
        )
        .route(
            "/:id/messages/:message_id/variables/recheck",
            post(recheck_message_variables),
        )
        .route("/:id/variables", get(list_variables).put(upsert_variable))
        .route("/:id/variables/:variable_id", delete(delete_variable))
        .route("/:id/summarize", post(summarize_chat))
        .route("/:id/summary", delete(delete_chat_summary))
        .route("/:id/summary/regenerate", post(regenerate_chat_summary))
        .route("/:id/stream", get(stream_chat))
}

async fn list_chats(State(state): State<AppState>) -> AppResult<Json<Vec<Chat>>> {
    Ok(Json(db::list_chats(&state.pool).await?))
}

async fn list_archived_chats(State(state): State<AppState>) -> AppResult<Json<Vec<Chat>>> {
    Ok(Json(db::list_archived_chats(&state.pool).await?))
}

async fn create_chat(
    State(state): State<AppState>,
    Json(payload): Json<ChatCreate>,
) -> AppResult<Json<Chat>> {
    Ok(Json(
        db::create_chat(&state.pool, payload.title, payload.character_id).await?,
    ))
}

async fn get_queue(State(state): State<AppState>) -> AppResult<Json<QueueStatus>> {
    let (running, queued) = db::list_queue(&state.pool).await?;
    Ok(Json(QueueStatus { running, queued }))
}

async fn get_chat(State(state): State<AppState>, Path(id): Path<i64>) -> AppResult<Json<Chat>> {
    Ok(Json(db::get_chat(&state.pool, id).await?))
}

async fn update_chat(
    State(state): State<AppState>,
    Path(id): Path<i64>,
    Json(payload): Json<ChatUpdate>,
) -> AppResult<Json<Chat>> {
    Ok(Json(db::update_chat(&state.pool, id, payload).await?))
}

async fn archive_chat(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> AppResult<Json<OkResponse>> {
    let _ = db::get_chat(&state.pool, id).await?;
    for job in db::list_active_jobs_for_chat(&state.pool, id).await? {
        let _ = state.queue.cancel_job(&state.pool, job.id).await;
    }
    db::archive_chat(&state.pool, id).await?;
    Ok(Json(OkResponse { ok: true }))
}

async fn restore_chat(State(state): State<AppState>, Path(id): Path<i64>) -> AppResult<Json<Chat>> {
    Ok(Json(db::restore_chat(&state.pool, id).await?))
}

async fn permanently_delete_chat(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> AppResult<Json<OkResponse>> {
    db::permanently_delete_chat(&state.pool, id).await?;
    Ok(Json(OkResponse { ok: true }))
}

async fn list_messages(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> AppResult<Json<Vec<Message>>> {
    crate::summarize::cleanup_stale_summary_markers(&state.pool, id).await?;
    Ok(Json(db::list_messages(&state.pool, id).await?))
}

async fn rewind_after_message(state: &AppState, chat_id: i64, message_id: i64) -> AppResult<usize> {
    let messages = db::list_messages(&state.pool, chat_id).await?;
    let Some(idx) = messages.iter().position(|m| m.id == message_id) else {
        return Err(AppError::not_found("Message not found"));
    };
    let after: Vec<Message> = messages[(idx + 1)..].to_vec();
    for message in &after {
        for job in db::list_active_jobs_for_message(&state.pool, message.id).await? {
            let _ = state.queue.cancel_job(&state.pool, job.id).await;
        }
    }
    let deleted = db::delete_messages_after(&state.pool, chat_id, message_id).await?;
    revert_variable_updates_from_messages(&state.pool, chat_id, &after).await?;
    Ok(deleted.len())
}

async fn send_message(
    State(state): State<AppState>,
    Path(id): Path<i64>,
    Json(payload): Json<SendMessageRequest>,
) -> AppResult<Json<Message>> {
    if payload.content.trim().is_empty() {
        return Err(AppError::bad_request("Message cannot be empty"));
    }
    let _ = db::get_settings(&state.pool).await?;
    db::insert_message(
        &state.pool,
        id,
        MessageRole::User,
        payload.content.trim().to_string(),
        false,
    )
    .await?;
    let assistant = db::insert_message(
        &state.pool,
        id,
        MessageRole::Assistant,
        String::new(),
        false,
    )
    .await?;
    let job = enqueue_generation(&state.pool, &state.queue, id, assistant.id).await?;
    db::touch_chat(&state.pool, id).await?;
    Ok(Json(Message {
        job_status: Some(job.status),
        ..assistant
    }))
}

async fn update_message(
    State(state): State<AppState>,
    Path((id, message_id)): Path<(i64, i64)>,
    Json(payload): Json<UpdateMessageRequest>,
) -> AppResult<Json<Message>> {
    if payload.content.trim().is_empty() {
        return Err(AppError::bad_request("Message cannot be empty"));
    }
    let message = db::get_message(&state.pool, id, message_id).await?;
    if message.is_summary {
        return Err(AppError::bad_request("Cannot edit summary messages"));
    }
    if message.role == MessageRole::System {
        return Err(AppError::bad_request("Cannot edit system messages"));
    }
    if payload.rewind {
        rewind_after_message(&state, id, message_id).await?;
    }
    db::update_message_content(&state.pool, message_id, payload.content.trim()).await?;
    if message.role == MessageRole::Assistant {
        db::clear_message_thoughts(&state.pool, message_id).await?;
    }
    db::touch_chat(&state.pool, id).await?;
    Ok(Json(db::get_message(&state.pool, id, message_id).await?))
}

async fn rewind_message(
    State(state): State<AppState>,
    Path((id, message_id)): Path<(i64, i64)>,
) -> AppResult<Json<OkResponse>> {
    let _ = db::get_message(&state.pool, id, message_id).await?;
    rewind_after_message(&state, id, message_id).await?;
    db::touch_chat(&state.pool, id).await?;
    Ok(Json(OkResponse { ok: true }))
}

async fn regenerate_message(
    State(state): State<AppState>,
    Path((id, message_id)): Path<(i64, i64)>,
    Json(payload): Json<RegenerateMessageRequest>,
) -> AppResult<Json<Message>> {
    let message = db::get_message(&state.pool, id, message_id).await?;
    if message.role != MessageRole::Assistant {
        return Err(AppError::bad_request(
            "Only assistant messages can be regenerated",
        ));
    }
    if message.is_summary {
        return Err(AppError::bad_request("Cannot regenerate summary messages"));
    }

    let variable_updates = message.variable_updates.clone();
    let is_last = db::is_last_message(&state.pool, id, message_id).await?;
    if !is_last || payload.rewind {
        rewind_after_message(&state, id, message_id).await?;
    }

    revert_message_variable_updates(&state.pool, id, message_id, &variable_updates).await?;

    for job in db::list_active_jobs_for_message(&state.pool, message_id).await? {
        state.queue.cancel_job(&state.pool, job.id).await?;
    }

    db::update_message_content(&state.pool, message_id, "").await?;
    db::clear_message_thoughts(&state.pool, message_id).await?;
    db::clear_message_variable_updates(&state.pool, message_id).await?;
    let job = enqueue_generation(&state.pool, &state.queue, id, message_id).await?;
    db::touch_chat(&state.pool, id).await?;
    let mut updated = db::get_message(&state.pool, id, message_id).await?;
    updated.job_status = Some(job.status);
    Ok(Json(updated))
}

async fn recheck_message_variables(
    State(state): State<AppState>,
    Path((id, message_id)): Path<(i64, i64)>,
) -> AppResult<Json<Job>> {
    let _ = db::get_message(&state.pool, id, message_id).await?;
    let settings = db::get_settings(&state.pool).await?;
    let job = state
        .queue
        .enqueue_variable_recheck(&state.pool, id, message_id, &settings)
        .await?;
    db::touch_chat(&state.pool, id).await?;
    Ok(Json(job))
}

async fn list_variables(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> AppResult<Json<Vec<ChatVariable>>> {
    Ok(Json(db::list_variables(&state.pool, id).await?))
}

async fn upsert_variable(
    State(state): State<AppState>,
    Path(id): Path<i64>,
    Json(payload): Json<ChatVariableUpdate>,
) -> AppResult<Json<ChatVariable>> {
    Ok(Json(
        db::upsert_variable_manual(&state.pool, id, payload).await?,
    ))
}

async fn delete_variable(
    State(state): State<AppState>,
    Path((id, variable_id)): Path<(i64, i64)>,
) -> AppResult<Json<OkResponse>> {
    let key = db::get_chat_variable(&state.pool, id, variable_id)
        .await?
        .key;
    db::delete_variable(&state.pool, id, variable_id).await?;
    strip_variable_key_from_chat_messages(&state.pool, id, &key).await?;
    db::touch_chat(&state.pool, id).await?;
    Ok(Json(OkResponse { ok: true }))
}

async fn summarize_chat(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> AppResult<Json<Job>> {
    let _ = db::get_chat(&state.pool, id).await?;
    let settings = db::get_settings(&state.pool).await?;
    let job = state
        .queue
        .enqueue_summarize(&state.pool, id, &settings)
        .await?;
    db::touch_chat(&state.pool, id).await?;
    Ok(Json(job))
}

async fn delete_chat_summary(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> AppResult<Json<OkResponse>> {
    crate::summarize::delete_chat_summary(&state.pool, id).await?;
    Ok(Json(OkResponse { ok: true }))
}

async fn regenerate_chat_summary(
    State(state): State<AppState>,
    Path(id): Path<i64>,
    Json(payload): Json<RegenerateSummaryRequest>,
) -> AppResult<Json<Job>> {
    let _ = db::get_chat(&state.pool, id).await?;
    let settings = db::get_settings(&state.pool).await?;
    let job = state
        .queue
        .enqueue_regenerate_summary(&state.pool, id, payload.marker_id, &settings)
        .await?;
    db::touch_chat(&state.pool, id).await?;
    Ok(Json(job))
}

async fn stream_chat(
    State(state): State<AppState>,
    Path(id): Path<i64>,
) -> AppResult<impl IntoResponse> {
    let _ = db::get_chat(&state.pool, id).await?;
    let pool = state.pool.clone();
    let interval = Duration::from_millis(state.sse_poll_interval_ms);

    let event_stream = stream! {
        let mut last_payload = String::new();
        loop {
            let chat = match db::get_chat(&pool, id).await {
                Ok(chat) => chat,
                Err(_) => {
                    yield Ok::<_, Infallible>(Event::default().event("error").data("{\"detail\":\"not found\"}"));
                    break;
                }
            };
            let messages = db::list_messages(&pool, id).await.unwrap_or_default();
            let active_job = db::get_active_job(&pool, id).await.ok().flatten();
            let has_active_job = active_job.is_some();
            let payload = serde_json::to_string(&ChatStreamPayload {
                chat,
                messages,
                active_job,
            }).unwrap_or_default();

            if payload != last_payload {
                last_payload = payload.clone();
                yield Ok(Event::default().data(payload));
            }

            if !has_active_job {
                yield Ok(Event::default().event("idle").data(format!("{{\"chat_id\":{id}}}")));
                break;
            }

            tokio::time::sleep(interval).await;
        }
    };

    Ok(Sse::new(event_stream).keep_alive(KeepAlive::default()))
}

#[cfg(test)]
mod stream_tests {
    use std::time::Duration;

    use axum::body::Body;
    use axum::http::{Request, StatusCode};
    use axum::Router;
    use dreamwell_types::{CharacterCreate, JobStatus, MessageRole};
    use http_body_util::BodyExt;
    use tower::ServiceExt;

    use crate::db;
    use crate::queue::JobQueue;
    use crate::routes::AppState;

    use super::stream_chat;

    async fn test_state(poll_ms: u64) -> (tempfile::TempDir, AppState) {
        let dir = tempfile::tempdir().expect("tempdir");
        let path = dir.path().join("test.db");
        let pool = db::connect(&format!("sqlite:{}", path.display()))
            .await
            .expect("connect");
        let queue = JobQueue::new(pool.clone());
        let state = AppState {
            pool,
            queue,
            sse_poll_interval_ms: poll_ms,
        };
        (dir, state)
    }

    async fn seed_chat(state: &AppState) -> (i64, i64) {
        let character = db::create_character(
            &state.pool,
            CharacterCreate {
                name: "Test".into(),
                description: String::new(),
                personality: String::new(),
                scenario: String::new(),
                first_message: String::new(),
                example_dialogue: String::new(),
                system_prompt: String::new(),
                avatar_url: None,
            },
        )
        .await
        .expect("character");
        let chat = db::create_chat(&state.pool, "Chat".into(), character.id)
            .await
            .expect("chat");
        let message = db::insert_message(
            &state.pool,
            chat.id,
            MessageRole::Assistant,
            String::new(),
            false,
        )
        .await
        .expect("message");
        (chat.id, message.id)
    }

    fn parse_sse_events(body: &str) -> Vec<(Option<String>, String)> {
        body.split("\n\n")
            .filter(|block| !block.trim().is_empty())
            .map(|block| {
                let mut event = None;
                let mut data = String::new();
                for line in block.lines() {
                    if let Some(name) = line.strip_prefix("event:") {
                        event = Some(name.trim().to_string());
                    } else if let Some(payload) = line.strip_prefix("data:") {
                        data = payload.trim().to_string();
                    }
                }
                (event, data)
            })
            .collect()
    }

    #[tokio::test]
    async fn stream_emits_data_then_idle_when_no_active_job() {
        let (_dir, state) = test_state(50).await;
        let (chat_id, _message_id) = seed_chat(&state).await;

        let app = Router::new()
            .route("/:id/stream", axum::routing::get(stream_chat))
            .with_state(state);

        let response = app
            .oneshot(
                Request::builder()
                    .uri(format!("/{chat_id}/stream"))
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);
        let body = response.into_body().collect().await.unwrap().to_bytes();
        let text = String::from_utf8_lossy(&body);
        let events = parse_sse_events(&text);
        assert!(!events.is_empty());
        assert!(events
            .iter()
            .any(|(event, data)| { event.is_none() && data.contains("\"messages\"") }));
        assert!(events
            .iter()
            .any(|(event, _)| event.as_deref() == Some("idle")));
    }

    #[tokio::test]
    async fn stream_emits_idle_after_job_completes() {
        let (_dir, state) = test_state(50).await;
        let (chat_id, message_id) = seed_chat(&state).await;
        let job = db::enqueue_job(&state.pool, chat_id, message_id)
            .await
            .expect("enqueue");
        sqlx::query("UPDATE generation_jobs SET status = 'running' WHERE id = ?1")
            .bind(job.id)
            .execute(&state.pool)
            .await
            .expect("mark running");

        let pool = state.pool.clone();
        let job_id = job.id;
        tokio::spawn(async move {
            tokio::time::sleep(Duration::from_millis(120)).await;
            db::complete_job(&pool, job_id, JobStatus::Completed, None)
                .await
                .expect("complete");
        });

        let app = Router::new()
            .route("/:id/stream", axum::routing::get(stream_chat))
            .with_state(state);

        let response = app
            .oneshot(
                Request::builder()
                    .uri(format!("/{chat_id}/stream"))
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);
        let body = response.into_body().collect().await.unwrap().to_bytes();
        let text = String::from_utf8_lossy(&body);
        let events = parse_sse_events(&text);
        assert!(events
            .iter()
            .any(|(event, _)| event.as_deref() == Some("idle")));
    }
}
